Mathlib/RingTheory/Valuation/ValuativeRel.lean

```diff
@@ -0,0 +1,582 @@
+This bundles a relation `rel : R → R → Prop` on `R` which mimics a
+preorder on `R` arising from a valuation.
+We introduce the notation `x ∣ᵥ y` for this relation.
+
+Recall that the equivalence class of a valuation is *completely* characterized by
+such a preorder. Thus, we can think of `ValuativeRel R` as a way of
+saying that `R` is endowed with an equivalence class of a valuation.
+-/
+
+noncomputable section
+
+/-- The class `[ValuativeRel R]` class introduces an operator `x ∣ᵥ y : Prop` for `x y : R`
+which is the natural relation arising from an equivalence class of a valuation on `R`.
+More precisely, if v is a valuation on R then the associated relation is `x ∣ᵥ y ↔ v x ≤ v y`.
+Use this class to talk about the case where `R` is equipped with an equivalence class
+of valuations. -/
+class ValuativeRel (R : Type*) [CommRing R] where
+  /-- The relation operator arising from `ValuativeRel`. -/
+  rel : R → R → Prop
+  rel_total (x y) : rel x y ∨ rel y x
+  rel_trans {z y x} : rel x y → rel y z → rel x z
+  rel_add {x y z} : rel x z → rel y z → rel (x + y) z
+  rel_mul_right {x y} (z) : rel x y → rel (x * z) (y * z)
+  rel_mul_cancel {x y z} : ¬ rel z 0 → rel (x * z) (y * z) → rel x y
+  not_rel_one_zero : ¬ rel 1 0
+
+@[inherit_doc] infix:50  " ≤ᵥ " => ValuativeRel.rel
+
+namespace Valuation
+
+variable {R Γ : Type*} [CommRing R] [LinearOrderedCommMonoidWithZero Γ]
+  (v : Valuation R Γ)
+
+/-- We say that a valuation `v` is `Compatible` if the relation `x ∣ᵥ y`
+is equivalent to `v x ≤ x y`. -/
+class Compatible [ValuativeRel R] where
+  rel_iff_le (x y : R) : x ≤ᵥ y ↔ v x ≤ v y
+
+end Valuation
+
+/-- A preorder on a ring is said to be "valuative" if it agrees with the
+valuative relation. -/
+class ValuativePreorder (R : Type*) [CommRing R] [ValuativeRel R] [Preorder R] where
+  dvd_iff_le (x y : R) : x ≤ᵥ y ↔ x ≤ y
+
+namespace ValuativeRel
+
+variable {R : Type*} [CommRing R] [ValuativeRel R]
+
+@[simp]
+lemma rel_refl (x : R) : x ≤ᵥ x := by
+  cases rel_total x x <;> assumption
+
+lemma rel_rfl {x : R} : x ≤ᵥ x :=
+  rel_refl x
+
+protected alias rel.refl := rel_refl
+
+protected alias rel.rfl := rel_rfl
+
+@[simp]
+theorem zero_rel (x : R) : 0 ≤ᵥ x := by
+  simpa using rel_mul_right x ((rel_total 0 1).resolve_right not_rel_one_zero)
+
+lemma rel_mul_left {x y : R} (z) : x ≤ᵥ y → (z * x) ≤ᵥ (z * y) := by
+  rw [mul_comm z x, mul_comm z y]
+  apply rel_mul_right
+
+instance : Trans (rel (R := R)) (rel (R := R)) (rel (R := R)) where
+  trans h1 h2 := rel_trans h1 h2
+
+lemma rel_mul {x x' y y' : R} : x ≤ᵥ y → x' ≤ᵥ y' → x * x' ≤ᵥ y * y' := by
+  intro h1 h2
+  calc x * x' ≤ᵥ x * y' := rel_mul_left _ h2
+    _ ≤ᵥ y * y' := rel_mul_right _ h1
+
+theorem rel_add_cases (x y : R) : x + y ≤ᵥ x ∨ x + y ≤ᵥ y :=
+  (rel_total y x).imp (fun h => rel_add .rfl h) (fun h => rel_add h .rfl)
+
+variable (R) in
+def unitSubmonoid : Submonoid R where
+  carrier := { x | ¬ x ≤ᵥ 0}
+  mul_mem' := by
+    intro x y hx hy
+    by_contra c
+    apply hy
```

Comment:
That should be
```
    contrapose! hy with c
```
but it doesn't work (maybe this is worth a bug report? Not sure)

Reply:
This works after a dsimp.